### PR TITLE
feat: add support for redis client

### PR DESCRIPTION
**Pull request by Antignote**
Adds support for existing Redis client to ISR cache handler

## Description

This PR enhances the `RedisCacheHandler` to accept an existing Redis client instance, providing more flexibility for applications that need to share a Redis connection or have custom Redis configurations.

### Changes

- **Updated `RedisCacheHandlerOptions.redis`** to accept:
  - An existing `Redis` client instance (ioredis)
  - A URL string (existing)
  - A `RedisOptions` object (existing)

- **Improved connection lifecycle management**: When an existing client is passed, the handler skips adding error listeners to avoid conflicts with the application's own error handling

- **Enhanced documentation** with examples showing both usage patterns:
  - Creating a new client from URL/config
  - Passing an existing client instance

### Benefits

- **Connection reuse**: Share a single Redis connection across multiple parts of your application
- **Custom configuration**: Use advanced Redis features (Cluster, Sentinel, etc.) by passing a pre-configured client
- **Better control**: Manage Redis connection lifecycle independently from the cache handler

### Example Usage

```javascript
import Redis from "ioredis";
import { RedisCacheHandler } from "@mrjasonroy/cache-components-cache-handler/handlers/redis";

const redisClient = new Redis({ host: "localhost", port: 6379 });

new RedisCacheHandler({ redis: redisClient });
```

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Chore (dependencies, CI, etc.)

## Testing

- [x] All existing tests pass (`pnpm test`)
- [ ] Added new tests for new functionality
- [x] Tested with memory cache handler
- [x] Tested with Redis cache handler (if applicable)
- [x] E2E tests pass (`pnpm test:e2e`)

## Checklist

- [x] My code follows the project's code style (`pnpm lint` passes)
- [x] I have run `pnpm format` to format my code
- [x] Type checking passes (`pnpm typecheck`)
- [x] I have updated the documentation accordingly
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Related Issues

N/A

### CODE REVIEW

File: packages/cache-handler/src/handlers/redis.ts [hunk 1]
```diff
@@ -77,24 +101,34 @@ export class RedisCacheHandler implements CacheHandler {
   private readonly tagPrefix: string;
   private readonly defaultTTL?: number;
   private readonly debug: boolean;
+  private didCreateClient = false;
 
   constructor(options: RedisCacheHandlerOptions = {}) {
     // Initialize Redis connection
-    if (typeof options.redis === "string") {
+    if (options.redis instanceof Redis) {
```
Comment (mrjasonroy) [inline packages/cache-handler/src/handlers/redis.ts]: 🔴 **Critical: `instanceof Redis` is fragile and will silently break across package versions.**

`instanceof` fails when the consumer's `ioredis` version differs from the one bundled here, or when bundlers (Webpack, esbuild) duplicate the module. When it fails, the code falls through to the `else` branch — calling `new Redis(options.redis)` with your client object as config. ioredis silently ignores unknown props and opens a new default connection to `localhost:6379`. No error thrown. You'll just be talking to the wrong Redis.

The sibling `data-cache/redis.ts` already solves this with a `RedisClient` interface (duck-typing). Recommend the same approach here:

```ts
function isRedisClient(value: unknown): value is Redis {
  return (
    typeof value === 'object' &&
    value !== null &&
    typeof (value as Redis).get === 'function' &&
    typeof (value as Redis).set === 'function' &&
    typeof (value as Redis).quit === 'function'
  );
}
```

**Review by mrjasonroy — CHANGES_REQUESTED** [review]
## Review: feat/add support for redis client

Good feature direction — passing an existing client is a real need for connection sharing, Sentinel/Cluster setups, and test mocking. The implementation is close but has two critical bugs and no tests.

**Verdict: Request Changes**

### Critical
- `instanceof Redis` is fragile and will silently fail across package versions/bundlers (see inline)
- The two `instanceof` checks at lines 108 and 127 can disagree, creating a worst-case state where a new phantom connection is opened AND no error listener is attached

### Blocking
- Zero tests — no `handlers/redis.test.ts` exists at all. This PR adds lifecycle-sensitive behavior (conditional `quit()`, conditional error listener) with no coverage.

### Questions for the author
1. Why `instanceof Redis` instead of a duck-type check or the `RedisClient` interface pattern used in `data-cache/redis.ts`? Deliberate divergence?
2. Is there a plan to add tests before merge?
3. Should `close()` be added to the `CacheHandler` interface? The behavior now differs depending on client ownership — composability matters.

File: packages/cache-handler/src/handlers/redis.ts [hunk 2]
```diff
@@ -77,24 +101,34 @@ export class RedisCacheHandler implements CacheHandler {
...
     this.tagPrefix = options.tagPrefix ?? "nextjs:tags:";
     this.defaultTTL = options.defaultTTL;
     this.debug = options.debug ?? false;
 
-    // Handle Redis connection errors
-    this.redis.on("error", (err) => {
-      console.error("[RedisCacheHandler] Redis connection error:", err);
-    });
+    // Handle Redis connection errors (only add listener if we created the client)
+    if (!(options.redis instanceof Redis)) {
```
Comment (mrjasonroy) [inline packages/cache-handler/src/handlers/redis.ts]: 🔴 **This check can disagree with line 108, creating a worst-case scenario.**

If `instanceof` is unreliable (see line 108 comment), both checks can fail simultaneously — meaning: the code creates a new client (line 108 falls through to `else`) AND skips the error listener here. That's a phantom connection with unhandled errors.

Use `this.didCreateClient` instead — it's already tracking this and is guaranteed to agree with the actual branching logic:

```ts
if (this.didCreateClient) {
  this.redis.on('error', (err) => { ... });
}
```

This is simpler, correct, and consistent with how `close()` at line 315 already does it.

File: packages/cache-handler/src/handlers/redis.ts [hunk 3]
```diff
@@ -77,24 +101,34 @@ export class RedisCacheHandler implements CacheHandler {
   private readonly tagPrefix: string;
   private readonly defaultTTL?: number;
   private readonly debug: boolean;
+  private didCreateClient = false;
```
Comment (mrjasonroy) [inline packages/cache-handler/src/handlers/redis.ts]: 🟡 **Nit: should be `readonly`.**

This flag is set once in the constructor and never mutated after. Mark it `readonly` for clarity and to let the type system enforce that:

```ts
private readonly didCreateClient: boolean;
```

Then assign it explicitly in each constructor branch (e.g. `this.didCreateClient = true` / `false`).

File: packages/cache-handler/src/handlers/redis.ts [hunk 4]
```diff
@@ -77,24 +101,34 @@ export class RedisCacheHandler implements CacheHandler {
...
     this.debug = options.debug ?? false;
 
-    // Handle Redis connection errors
-    this.redis.on("error", (err) => {
-      console.error("[RedisCacheHandler] Redis connection error:", err);
-    });
+    // Handle Redis connection errors (only add listener if we created the client)
+    if (!(options.redis instanceof Redis)) {
+      this.redis.on("error", (err) => {
+        console.error("[RedisCacheHandler] Redis connection error:", err);
```
Comment (mrjasonroy) [inline packages/cache-handler/src/handlers/redis.ts:139]: 🟡 **Foot-gun: if the consumer also forgot to add an error listener, Node.js will crash.**

Multiple `'error'` listeners on the same EventEmitter are fine — they're called in order. A safer approach: always attach one, but check first so we don't add a duplicate logger:

```ts
if (this.redis.listenerCount('error') === 0) {
  this.redis.on('error', (err) => {
    console.error('[RedisCacheHandler] Redis connection error:', err);
  });
}
```

At minimum, the JSDoc for the shared-client option should document that the caller is responsible for error handling.

**Review by mrjasonroy — COMMENTED** [review]
Thanks for this contribution, @Antignote! Accepting an existing Redis client instance is a great feature for connection reuse and advanced configs (Cluster, Sentinel, etc.). The `didCreateClient` lifecycle pattern is well done.

A few things to address before we can merge:

### 1. `instanceof Redis` fragility across package boundaries
If the user's app has a different ioredis version than this package's dependency, `instanceof` will return `false` (different constructors from different `node_modules` copies). A duck-typing check would be more robust:

```typescript
function isRedisInstance(value: unknown): value is Redis {
  return (
    typeof value === "object" &&
    value !== null &&
    typeof (value as Redis).get === "function" &&
    typeof (value as Redis).set === "function" &&
    typeof (value as Redis).del === "function"
  );
}
```

Or at minimum, document in the JSDoc that the ioredis version must match.

### 2. Inconsistent condition for error listener
The error listener guard (line ~125) duplicates the `instanceof` check:

```typescript
if (!(options.redis instanceof Redis)) {
  this.redis.on("error", ...);
}
```

This should use the `didCreateClient` flag you already have, for consistency and to avoid the `instanceof` issue:

```typescript
if (this.didCreateClient) {
  this.redis.on("error", (err) => { ... });
}
```

### 3. No tests
Would like to see at least:
- Passing an existing client instance and verifying get/set works
- `close()` does **not** call `quit()` on a shared client
- `close()` **does** call `quit()` on an internally-created client

### 4. Needs rebase
This branch is from Jan 14 and main has had 14+ commits since. Will need a rebase to merge cleanly.

---

Overall the approach is solid — just needs these cleanups. Let us know if you'd like help with any of them!

**Comment by mrjasonroy** [comment]
Rebased onto main and pushed a fix commit addressing the review feedback:

1. **Duck typing instead of `instanceof Redis`** - `instanceof` can break across package versions or bundler setups. Now checks for `get`/`set`/`del` methods instead.
2. **Consistent `didCreateClient` usage** - Error listener condition now uses `this.didCreateClient` instead of a second `instanceof` check.
3. **Added 5 tests** - Covers existing client detection, error listener behavior, and `close()` lifecycle management.

All 95 tests pass. Ready to merge once CI is green.

**Comment by johannes-wtfoxtrot** [comment]
@mrjasonroy CI is green - is this ready to merge? It'd be great to be able to pass a redis client for sentinel support.

**Review by mrjasonroy — DISMISSED** [review]
Thanks for sticking with this, @Antignote — and thanks to @johannes-wtfoxtrot for the nudge. All the earlier review points are now resolved:

- ✅ **Duck typing instead of `instanceof Redis`** — robust across package versions/bundlers
- ✅ **Consistent lifecycle via `didCreateClient`** — no more disagreeing `instanceof` checks
- ✅ **Tests added** — existing-client get/set, conditional `close()`/`quit()`, and error-listener behavior

I pushed one follow-up commit on top to harden two edge cases:
1. The duck-type check now also requires `pipeline` and `on`, so a plain `{get,set,del}` object isn't mistaken for a Redis client.
2. A defensive `error` listener is attached to an external client only when it has none — an ioredis EventEmitter with no error listener will crash the process on a connection error.

All 97 tests pass locally. Approving — merging once CI is green.

File: packages/cache-handler/src/handlers/redis.ts (hunk 3 above)
Comment (mrjasonroy) [reply packages/cache-handler/src/handlers/redis.ts]: Done in 0e76ec3 — `didCreateClient` is now `readonly`.

File: packages/cache-handler/src/handlers/redis.ts (hunk 4 above)
Comment (mrjasonroy) [reply packages/cache-handler/src/handlers/redis.ts:139]: Addressed in 9649518 — for external clients we now attach a defensive `error` listener when the client has none (`listenerCount("error") === 0`), so a consumer who forgot one won't crash the process. Covered by a new test.

**Review by mrjasonroy — APPROVED** [review]
Re-approving after the readonly cleanup commit. All feedback addressed, threads resolved.
